src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -54,18 +54,24 @@ private void refreshCacheItem(final MessageCacheItem cacheItem, Iterator<DataSou
 
 		long timestampOld = cacheItem.getTimestamp();
 		DataSourceEnum dataSource = msgSourceQueueIter.next();
+		String localeOrig = dto.getLocale();
+		if (dataSource.equals(DataSourceEnum.VIP) && dto.getLocale().equals("source")) {
+			dto.setLocale("latest");
+		}
 		dataSource.createMessageOpt(dto).getComponentMessages(cacheItem);
 		long timestamp = cacheItem.getTimestamp();
 		if (timestampOld == timestamp) {
 			logger.debug(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED, dto.getComponent(), dto.getLocale(), dataSource.toString()));
 		}
+		dto.setLocale(localeOrig);
 
-		// If timestamp is not 0, it means that cacheItem is in the cache already.
-		// Otherwise, try the next dataSource in the queue.
-		if (timestamp == 0) {
+		// If timestamp is 0, it means that cacheItem not yet in cache. So try the next data source.
+		// If locale is "source", use messages from DatasourceEnum.Bundle data store if it exists.
+		if (timestamp == 0 || dto.getLocale().equals("source")) {
```

Comment:
If getting latest feom service successfully, is it still necessary to get from local?

Reply:
I am not so sure, actually. Is there a use case where a string has been updated only in local and not yet in remote? If yes, should the client use the local message or wait until message in service is updated?

Reply:
If you think locale take priority over remote, then you should get from local first, then remote. Don't need to get from service first then get from local.

Reply:
I have studied existing use cases. As a conclusion, let's keep it consistent with the behavior for all locales: Pull from remote when remote is available, from local if remote is not available. I'll remove the change from line 70.